File: frontend/src/app/features/customers/domain/customer.ts
import { CustomerDirectoryResponse } from '../../../api/generated/model/customerDirectoryResponse';
import { CustomerProfileResponse } from '../../../api/generated/model/customerProfileResponse';
import { CustomerVerificationResponse } from '../../../api/generated/model/customerVerificationResponse';
import { VerificationFactor } from '../../../api/generated/model/verificationFactor';
import { parseUtcTimestamp } from '../../tickets/domain/ticket';

export interface CustomerSummary {
  readonly id: number;
  readonly customerNumber: string;
  readonly name: string;
  readonly email: string | null;
  readonly phone: string | null;
  readonly city: string | null;
  readonly country: string | null;
  readonly isActive: boolean;
}

export interface Customer extends CustomerSummary {
  readonly firstName: string;
  readonly lastName: string;
  readonly dateOfBirth: string | null;
  readonly street: string | null;
  readonly houseNumber: string | null;
  readonly postalCode: string | null;
  readonly createdAt: Date;
  readonly updatedAt: Date;
}

export interface CustomerPage {
  readonly items: readonly CustomerSummary[];
  readonly page: number;
  readonly pageSize: number;
  readonly total: number;
  readonly totalPages: number;
}

export interface CustomerVerification {
  readonly id: number;
  readonly customerId: number;
  readonly verifiedByUserId: number;
  readonly factors: readonly VerificationFactor[];
  readonly verifiedAt: Date;
  readonly expiresAt: Date;
}

export function mapCustomerSummary(dto: CustomerDirectoryResponse): CustomerSummary {
  return {
    id: dto.id,
    customerNumber: dto.customer_number,
    name: `${dto.first_name} ${dto.last_name}`.trim(),
    email: dto.email,
    phone: dto.phone,
    city: dto.city,
    country: dto.country,
    isActive: dto.is_active,
  };
}

export function mapCustomer(dto: CustomerProfileResponse): Customer {
  return {
    ...mapCustomerSummary(dto),
    firstName: dto.first_name,
    lastName: dto.last_name,
    dateOfBirth: dto.date_of_birth,
    street: dto.street,
    houseNumber: dto.house_number,
    postalCode: dto.postal_code,
    createdAt: parseUtcTimestamp(dto.created_at),
    updatedAt: parseUtcTimestamp(dto.updated_at),
  };
}

export function mapVerification(dto: CustomerVerificationResponse): CustomerVerification {
  return {
    id: dto.id,
    customerId: dto.customer_id,
    verifiedByUserId: dto.verified_by_user_id,
    factors: dto.factors,
    verifiedAt: parseUtcTimestamp(dto.verified_at),
    expiresAt: parseUtcTimestamp(dto.expires_at),
  };
}

File: backend/app/models/customer.py
from datetime import date, datetime
from typing import TYPE_CHECKING

from sqlalchemy import Boolean, CheckConstraint, Index, String
from sqlalchemy.dialects.mysql import BIGINT, CHAR, DATETIME
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database.base import Base

if TYPE_CHECKING:
    from app.models.customer_email_delivery import CustomerEmailDelivery
    from app.models.customer_verification import CustomerVerification
    from app.models.ticket import Ticket


class Customer(Base):
    __tablename__ = "customers"

    __table_args__ = (
        CheckConstraint(
            "email IS NOT NULL OR phone IS NOT NULL",
            name="contact_present",
        ),
        Index("ix_customers_email", "email"),
        Index("ix_customers_phone", "phone"),
        Index(
            "ix_customers_active_last_first_id",
            "is_active",
            "last_name",
            "first_name",
            "id",
        ),
        Index(
            "ix_customers_active_first_last_id",
            "is_active",
            "first_name",
            "last_name",
            "id",
        ),
    )

    id: Mapped[int] = mapped_column(
        BIGINT(unsigned=True), primary_key=True, autoincrement=True
    )
    customer_number: Mapped[str] = mapped_column(
        String(20), nullable=False, unique=True
    )
    first_name: Mapped[str] = mapped_column(String(100), nullable=False)
    last_name: Mapped[str] = mapped_column(String(100), nullable=False)
    date_of_birth: Mapped[date | None] = mapped_column(nullable=True)
    email: Mapped[str | None] = mapped_column(
        String(254).with_variant(String(254, collation="utf8mb4_bin"), "mysql"),
        nullable=True,
    )
    phone: Mapped[str | None] = mapped_column(String(16), nullable=True)
    street: Mapped[str | None] = mapped_column(String(150), nullable=True)
    house_number: Mapped[str | None] = mapped_column(String(30), nullable=True)
    postal_code: Mapped[str | None] = mapped_column(String(20), nullable=True)
    city: Mapped[str | None] = mapped_column(String(100), nullable=True)
    country: Mapped[str | None] = mapped_column(CHAR(2), nullable=True)
    is_active: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=True, server_default="1"
    )
    created_at: Mapped[datetime] = mapped_column(DATETIME(fsp=6), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DATETIME(fsp=6), nullable=False)

    verifications: Mapped[list["CustomerVerification"]] = relationship(
        back_populates="customer"
    )
    tickets: Mapped[list["Ticket"]] = relationship(
        back_populates="customer",
        foreign_keys="Ticket.customer_id",
    )
    email_deliveries: Mapped[list["CustomerEmailDelivery"]] = relationship(
        back_populates="customer",
        foreign_keys="CustomerEmailDelivery.customer_id",
    )
